lib/perl/Genome/SoftwareResult/User.pm

```diff
@@ -39,5 +43,86 @@ class Genome::SoftwareResult::User {
     doc => 'links a software result to other entities which depend on it',
 };
 
+sub with_registered_users {
+    my $class = shift;
+    my %params = Params::Validate::validate(
+        @_, {
+            users => {
+                type      => HASHREF,
+                optional  => 0,
+                callbacks => {
+                    'must contain sponsor and requestor' => \&_validate_user_hash
+                },
+            },
+            callback => {
+                type     => CODEREF,
+                optional => 0,
+            }
+        }
+    );
+
+    my $user_hash = delete $params{users};
+    my $sr_callback = delete $params{callback};
+
+    my ($software_result, $newly_created) = $sr_callback->(%params);
+    return unless $software_result;
+
+    $class->_register_users($software_result, $user_hash, $newly_created);
+
+    return $software_result;
+}
+
+sub _validate_user_hash {
+    my $user_hash = shift;
+
+    for my $type (qw( sponsor requestor )) {
+        my $obj = $user_hash->{$type};
+        return 0 unless($obj);
+        return 0 unless(any { UNIVERSAL::isa($obj, $_) } _valid_types($type));
+    }
+
+    return 1;
+}
+
+sub _register_users {
+    my $class = shift;
+    my $software_result = shift;
+    my $user_hash = shift;
+    my $newly_created = shift;
+
+    my %user_hash = %$user_hash;
+
+    my $requestor = delete $user_hash{requestor};
+    my $label = $newly_created ? 'created' : 'shortcut';
+    $user_hash{$label} = $requestor;
+
+    while(my ($label, $object) = each %user_hash) {
+        my %params = (
+            label           => $label,
+            user            => $object,
+            software_result => $software_result,
+        );
+       $class->get(%params) || $class->create(%params);
+    }
+}
+
+
+use constant VALID_TYPES => {
+    requestor => ['Genome::Process', 'Genome::Model::Build'],
+    sponsor   => ['Genome::Sys::User', 'Genome::Config::AnalysisProject'],
+};
+
+sub _valid_types { return @{VALID_TYPES->{shift()}} }
+
+sub user_hash_for_build {
+    my $class = shift;
+    my $build = shift;
+
+    return {
+        requestor => $build,
+        sponsor   => $build->model->analysis_projects // Genome::Sys::User->get(username => $build->model->run_as),
```

Comment:
I think `run_as` better captures the intent.  If, for example, someone runs a build of a model for debugging purposes, they're not a stakeholder in that result.